### ISSUE TITLE
Refinement lost inside map

### ISSUE THREAD

**Issue by maxsalven** [facebook/flow#6464]
Type refinements seem to be lost inside an Array.map call:

```
/* @flow */

type Action =
  | {
      flag: true,
      other: true,
    }
  | {
      flag: false,
    };

function reducer(action: Action): any {
  switch (action.flag) {
    case true:
      // return action.other; // This is ok
      return [1].map(item => action.other); // This is not
    default:
      return;
  }
}
```

https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgCCAxhgJZwB2YAvKmGAD5gDe9DHsAhgOYBcYDACcArngA07DmDgYAFniEDhYydIC+7Zm2kNu-MFC4wAzhKlh1AbnRQRlUhWpC8AExHFFACi6OqAknIqAEoBLkosVnYTBDIMYjkwHz9KADp9YKjpYi4zQVE8PgsGYGAwFwwRIWpfILTZBSErMFKwABU5MhMwLpkAa2LyvErqsABtAEYAXVSAWy4cLzi8WdoAPjBap1SGxWDm1o7e3spZC1c8IxEYDCLdIZHKGwZNTSA

**Comment by wchargin** [comment]
This is by design.

Flow only knows that the type of `map` is (roughly)

```js
// in class Array<T>...
map<U>(callbackfn: (T) => U): Array<U>;
```

You and I know that `map` does not mutate any state, and only invokes
the callback within the execution of `map` (i.e., does not save it for
later). But I could write a function `weirdMap` that does the following:

```js
Array.prototype.weirdMap = function weirdMap(callback) {
  const result = [];
  // Behave just like `map`...
  for (let i = 0; i < this.length; i++) {
    result.push(callback(this[i]));
  }
  // ...but then do something else:
  if (this.length > 0) {
    const first = this[0];
    window.setTimeout(() => {
      callback(first);
    }, 1000);
  }
  return result;
}
```

Now, if we write

```js
function go(action: Action): string[] | void {
  if (action.flag) {
    return [1].weirdMap(item => action.other.toString());
  }
}
go(anAction);
anAction = {flag: false};
```

then we will observe a runtime error: `reducer` returns successfully,
but one second later the callback fires again, and by that point
`anAction.other` is `undefined`.

Clearly, the distinction between `map` and `weirdMap` is important. But
the critical point is that that the type of `weirdMap` is the same as
the type of `map`. So, as far as Flow is concerned, `Array`’s `map`
could really be `weirdMap`. Therefore, the access to `action.other` is
invalid, because the refinement that `action.flag` is `true` is no
longer known.

Anyway, long story short: as it says in the docs, [function calls
invalidate refinements][1].

[1]: https://flow.org/en/docs/lang/refinements/#toc-refinement-invalidations

**Comment by Persola** [comment]
@wchargin Thanks for the explanation. Is it correct to say that the function call invalidates the refinements _inside the scope of the callback_?

Could someone compare this situation to [this version](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgCCAxhgJZwB2YAvKmGAD5gDe9DHsAhgOYBcYDACcArngA07DmDgYAFniEDhYydIC+7Zm2kNu-MFC4wAzhKlh1AbnRQRlUhWpC8AExHFFACi6OqAknIqAEoBLkosVnYTBDIMYjkwHz9KADp9YKjpYi4zQVE8PgsGAG1fIMoAXVSAWy4cLzi8GtoAPjAARmCbXTAXDBEhanKnVNkFIR6OVzwjERgMIt7+wcopzU0gA) which doesn't error? The change:
```
    case true:
      [action].map(item => 1);
      return action.other;
```
If Flow is agnostic as to whether `map` has side effects, the call to `action.other` could be invalid because `action` has been mutated by then, no?

**Comment by jussi-kalliokoski** [comment]
Refinements are also lost just when you create a new function (flow 0.82.0):

```
/* @flow */

interface X {
  y: string;
}

function foo() : () => void {
  let x: X | void;
  x = { y: "foo" };
  x.y = "boo"; // no error
  return () => {
    x.y = "zoo"; // Cannot assign `"zoo"` to `x.y` because property `y` is missing in undefined [1].
  };
}
```

If this is by design, it brings the question whether or not the design is good, as in the above case there is nothing that can cause `x` to become `void` again.

**Comment by SamChou19815** [comment]
Now both examples from OP work.
